lib/runtime-core/src/vm.rs

```diff
@@ -13,6 +13,25 @@ use std::{ffi::c_void, mem, ptr};
 #[derive(Debug)]
 #[repr(C)]
 pub struct Ctx {
+    // `internal` must be the first field of `Ctx`.
+    pub internal: InternalCtx,
```

Comment:
Can you make this private? You can safely cast from `Ctx` to `InternalCtx` even with this being private.